src/server/handlers/consents/{ID}/generateChallenge.ts

```diff
@@ -0,0 +1,140 @@
+ */
+export async function generateChallengeAndPutConsentId (
+  request: Request,
+  id: string
+): Promise<void> {
+  try {
+    // Fetch consent from database using ID
+    let consent: Consent
+    try {
+      consent = await consentDB.retrieve(id)
+    } catch (error) {
+      Logger.push(error)
+      Logger.error('Error in retrieving consent')
+
+      // If consent cannot be retrieved using given ID, send PUT ...error back
+      // TODO: Error Handling dealt with in future ticket #355
+      throw (new Error('NotImplementedYetError'))
+    }
+
+    if (!isConsentRequestInitiatedByValidSource(request, consent)) {
+      // TODO: Error Handling dealt with in future ticket #355
+      // send PUT ...error back
+      throw (new Error('NotImplementedYetError'))
+    }
+
+    // If there is no pre-existing challenge for the consent id
+    // Generate one and update database
+    if (!consent.credentialChallenge) {
+      // Challenge generation
+      const challengeValue = await challenge.generate()
+
+      // Updating credentials with generated challenge
+      const credential: ConsentCredential = {
+        credentialChallenge: challengeValue,
+        credentialStatus: 'PENDING',
+        credentialType: 'FIDO',
+        credentialPayload: null
+      }
+
+      consent = await updateConsentCredential(consent, credential)
+    } else if (consent.credentialStatus === 'ACTIVE') {
+      // TODO: Error handling here - dealt with in #355
+      Logger.error('ACTIVE credential consent has requested challenge')
+      throw (new Error('NotImplementedYetError'))
+    }
+
+    // Retrieve Scopes
+    const scopesRetrieved: Scope[] = await scopeDB.retrieveAll(id)
+    const scopes = convertScopesToExternal(scopesRetrieved)
+
+    // Outgoing call to PUT consents/{ID}
+
+    // Build Request Body
+    const requestBody = await generatePutConsentsRequest(consent, scopes)
+    // Use sdk-standard-components library to send request
+    await thirdPartyRequest.putConsents(
+      consent.id, requestBody, request.headers[Enum.Http.Headers.FSPIOP.SOURCE])
+  } catch (error) {
+    Logger.push(error)
+    Logger.error(`Outgoing call NOT made to PUT consent/${id}`)
+    // TODO: Decide on error handling HERE - dealt with in future ticket #355
+    throw error
+  }
+}
+
+/** The HTTP request `POST /consents/{ID}/generateChallenge` is used to create a
+ * credential for the given Consent object. The `{ID}` in the URI should
+ * contain the `{ID}` that was used in the `POST /consents`.
+ * Called by a `PISP`to request a challenge from the `auth-service`, which
+ * will be returned to the PISP via `PUT /consents/{ID}`
+ */
+export async function post (
+  request: Request, h: ResponseToolkit): Promise<ResponseObject> {
+  const id = request.params.id
+
+  // Asynchronously deals with validating request,
+  //  generating challenge, updating consent db
+  //  and making outgoing PUT consent/{ID} call
+  generateChallengeAndPutConsentId(request, id)
```

Comment:
```suggestion
  generateChallengeAndPutConsent(request, id)
```

Should we use a slightly different name? Or are we putting an ID?

Reply:
ah we're making a call to PUT consent/{ID} 
but either works?

Reply:
done